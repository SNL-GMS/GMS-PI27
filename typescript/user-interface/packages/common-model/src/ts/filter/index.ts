export * from './filter-util';
export * from './types';
