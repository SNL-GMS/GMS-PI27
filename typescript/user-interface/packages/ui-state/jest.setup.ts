/* eslint-disable max-classes-per-file */
/* eslint-disable import/first */
/* eslint-disable import/no-extraneous-dependencies */
import 'fake-indexeddb/auto';

import type { FilterDefinition } from '@gms/common-model/lib/filter/types';
import crypto from 'crypto';
import * as util from 'util';

import {
  channelSegmentWithSamples,
  valuesAsNumbers
} from './__tests__/__data__/ui-channel-segments/ui-channel-segment-data';

// This is to fix an issue with test leaks related to createStateSyncMiddleware() in jsdom
process.env.GMS_DISABLE_REDUX_STATE_SYNC = 'true';

process.env.GMS_DISABLE_PRE_CACHE_CHANNEL_SEGMENT_BY_CHANNEL = 'false';

Object.defineProperty(global, 'crypto', {
  value: {
    subtle: {
      digest: async (algorithm: AlgorithmIdentifier, msg: Uint8Array) =>
        (crypto.webcrypto as any).subtle.digest(algorithm, msg) as Promise<any>
    }
  }
});

Object.defineProperty(window, 'TextEncoder', {
  writable: true,
  value: util.TextEncoder
});
Object.defineProperty(window, 'TextDecoder', {
  writable: true,
  value: util.TextDecoder
});
Object.defineProperty(global, 'TextEncoder', {
  writable: true,
  value: util.TextEncoder
});
Object.defineProperty(global, 'TextDecoder', {
  writable: true,
  value: util.TextDecoder
});

import type { CommonTypes, WaveformTypes } from '@gms/common-model';
import { TimeseriesType } from '@gms/common-model/lib/channel-segment';
import type { GlobalWithFetchMock } from 'jest-fetch-mock';

const customGlobal: GlobalWithFetchMock = global as unknown as GlobalWithFetchMock;
// eslint-disable-next-line @typescript-eslint/no-require-imports
customGlobal.fetch = require('jest-fetch-mock');

customGlobal.fetchMock = customGlobal.fetch;

// eslint-disable-next-line @typescript-eslint/no-require-imports
require('jest-canvas-mock');

const globalAny: any = global;

// eslint-disable-next-line @typescript-eslint/no-require-imports
globalAny.fetch = require('jest-fetch-mock');

globalAny.TextEncoder = util.TextEncoder;

globalAny.window = window;
globalAny.console = {
  ...globalAny.console,
  warn: jest.fn()
};

class Worker {
  public url: string;

  public onmessage: (message: Record<string, unknown>, transfer?: Transferable[]) => void;

  public constructor(stringUrl: string) {
    this.url = stringUrl;
    this.onmessage = () => {
      // do nothing
    };
  }

  public postMessage(message: Record<string, unknown>, transfer?: Transferable[]) {
    this.onmessage(message, transfer);
  }
}
globalAny.window.Worker = Worker;

class MessagePort {
  public url: string;

  public onmessage: (message: Record<string, unknown>, transfer?: Transferable[]) => void;

  public postMessage(message: Record<string, unknown>, transfer?: Transferable[]) {
    this.onmessage(message, transfer);
  }
}

class SharedWorker {
  public url: string;

  public port = new MessagePort();

  public constructor(stringUrl: string) {
    this.url = stringUrl;

    this.port.onmessage = () => {
      // do nothing
    };
  }
}
globalAny.window.SharedWorker = SharedWorker;

// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-require-imports
require('jest-fetch-mock').enableMocks();

jest.mock('@gms/ui-wasm', () => {
  const isDesigned = jest.requireActual('@gms/ui-wasm/lib/gms-interop/is-designed');

  const filter = async (
    filterDefinition: FilterDefinition,
    data: Float64Array
  ): Promise<Float64Array> => Promise.resolve(data);

  const design = async (filterDefinition: FilterDefinition): Promise<FilterDefinition> =>
    Promise.resolve(filterDefinition);

  const getBoundsForPositionBuffer = async (): Promise<{
    min: number;
    minSecs: number;
    max: number;
    maxSecs: number;
  }> => Promise.resolve({ min: 0, minSecs: 0, max: 10, maxSecs: 10 });

  const waveformNorth: WaveformTypes.Waveform = {
    samples: new Float64Array(valuesAsNumbers),
    type: TimeseriesType.WAVEFORM,
    startTime: 0,
    endTime: 10,
    sampleRateHz: 40,
    sampleCount: valuesAsNumbers.length
  };

  const waveformEast: WaveformTypes.Waveform = {
    samples: new Float64Array(valuesAsNumbers),
    type: TimeseriesType.WAVEFORM,
    startTime: 11,
    endTime: 20,
    sampleRateHz: 40,
    sampleCount: valuesAsNumbers.length
  };

  const maskAndRotate2d = async (): Promise<
    CommonTypes.TimeseriesWithMissingInputChannels<WaveformTypes.Waveform>[]
  > =>
    Promise.resolve([
      {
        timeseries: [waveformNorth],
        missingInputChannels: []
      },
      {
        timeseries: [waveformEast],
        missingInputChannels: []
      }
    ]);

  return {
    design,
    filter,
    maskAndRotate2d,
    getBoundsForPositionBuffer,
    isDesigned: isDesigned.isDesigned,
    // eslint-disable-next-line @typescript-eslint/require-await
    maskAndBeamWaveforms: jest.fn(async () => channelSegmentWithSamples)
  };
});

// eslint-disable-next-line @typescript-eslint/no-magic-numbers
Date.now = jest.fn().mockReturnValue(250);
