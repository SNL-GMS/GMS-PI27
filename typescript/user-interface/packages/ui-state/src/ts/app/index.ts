export * from './api';
export * from './history';
export * from './hooks';
export * from './loader';
export * from './processing';
export * from './query';
export * from './redux-provider';
export * from './sanitization';
export * from './state';
export * from './store';
export * from './subscription';
export * from './util';
