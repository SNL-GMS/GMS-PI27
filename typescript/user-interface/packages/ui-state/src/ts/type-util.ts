import type { ChannelSegmentTypes } from '@gms/common-model';

import type { UiChannelSegment } from './types';

/**
 * Test to compare two {@link ChannelSegmentTypes.ChannelSegmentDescriptor}s to see if they are equal.
 *
 * @param a {@link ChannelSegmentTypes.ChannelSegmentDescriptor} to compare
 * @param b {@link ChannelSegmentTypes.ChannelSegmentDescriptor} to compare
 * @returns true if both {@link ChannelSegmentTypes.ChannelSegmentDescriptor} are equal
 */
export const areChannelSegmentDescriptorsEqual = (
  a: ChannelSegmentTypes.ChannelSegmentDescriptor,
  b: ChannelSegmentTypes.ChannelSegmentDescriptor
): boolean =>
  a.startTime === b.startTime && a.endTime === b.endTime && a.channel.name === b.channel.name;

/**
 * Test to compare two {@link UiChannelSegment}s to see if they have
 * the same (equal) {@link ChannelSegmentTypes.ChannelSegmentDescriptor}.
 *
 * @param a {@link UiChannelSegment} to compare
 * @param b {@link UiChannelSegment}  to compare
 * @returns true if both segments have the same {@link ChannelSegmentTypes.ChannelSegmentDescriptor}
 */
export function areUiChannelSegmentChannelSegmentDescriptorsEqual<
  T extends ChannelSegmentTypes.Timeseries
>(a: UiChannelSegment<T>, b: UiChannelSegment<T>) {
  return areChannelSegmentDescriptorsEqual(a.channelSegmentDescriptor, b.channelSegmentDescriptor);
}
