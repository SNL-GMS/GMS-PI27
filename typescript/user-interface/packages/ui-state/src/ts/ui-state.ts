export * from './app';
export * from './imperative-context-menu';
export * from './types';
export * from './workers';
