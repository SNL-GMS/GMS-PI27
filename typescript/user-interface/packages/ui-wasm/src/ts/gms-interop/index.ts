export * from './design';
export * from './filter';
export * from './fk/fk';
export * from './is-designed';
export * from './rotation/rotation';
