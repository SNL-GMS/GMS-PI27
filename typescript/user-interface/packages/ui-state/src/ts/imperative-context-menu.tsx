import type { ImperativeContextMenuProps } from '@gms/ui-core-components';
import { showImperativeContextMenu } from '@gms/ui-core-components';
import React from 'react';
import { Provider } from 'react-redux';

import { getStore } from './app/store';

/**
 * Show imperative context menu.
 *
 * Wraps the content with a Redux {@link Provider} and injects the Reject store
 * so that the store is available within the context of the context menu.
 *
 * @see {@link showImperativeContextMenu}
 */
export function showImperativeReduxContextMenu(props: ImperativeContextMenuProps) {
  const { content } = props;
  const propsWithRedux: ImperativeContextMenuProps = {
    ...props,
    content: <Provider store={getStore()}>{content}</Provider>
  };
  showImperativeContextMenu(propsWithRedux);
}
