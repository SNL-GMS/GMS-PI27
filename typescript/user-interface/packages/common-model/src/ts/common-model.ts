import * as ArrayUtil from './array-util';
import * as BeamformingTemplateTypes from './beamforming-templates/types';
import * as ChannelSegmentTypes from './channel-segment';
import * as ColorTypes from './color/types';
import * as CommonTypes from './common';
import * as Displays from './displays/types';
import * as Endpoints from './endpoints/types';
import * as EventTypes from './event';
// required for legacy components
import * as LegacyEventTypes from './event/legacy';
import * as FacetedTypes from './faceted';
import * as FilterUtil from './filter/filter-util';
import * as FilterTypes from './filter/types';
import * as FkTypes from './fk';
import * as ProcessingMaskDefinitionTypes from './processing-mask-definitions/types';
import * as QcSegmentTypes from './qc-segment';
import * as RotationTypes from './rotation/types';
import * as SignalDetectionTypes from './signal-detection';
import * as ChannelTypes from './station-definitions/channel-definitions';
import * as ResponseTypes from './station-definitions/response-definitions';
import * as StationTypes from './station-definitions/station-definitions/station-definitions';
import * as ProcessingStationTypes from './station-processing/types';
import * as ReferenceStationTypes from './station-reference/types';
import * as TimeTypes from './time';
import * as TypeUtil from './type-util/type-util';
import * as ConfigurationTypes from './ui-configuration/types';
import * as UserProfileTypes from './user-profile/types';
import * as WaveformTypes from './waveform/types';
import * as WaveformUtil from './waveform/util';
import * as WorkflowTypes from './workflow/types';

export * from './faceted/utils';

export {
  ArrayUtil,
  BeamformingTemplateTypes,
  ChannelSegmentTypes,
  ChannelTypes,
  ColorTypes,
  CommonTypes,
  ConfigurationTypes,
  Displays,
  Endpoints,
  EventTypes,
  FacetedTypes,
  FilterTypes,
  FilterUtil,
  FkTypes,
  LegacyEventTypes,
  ProcessingMaskDefinitionTypes,
  ProcessingStationTypes,
  QcSegmentTypes,
  ReferenceStationTypes,
  ResponseTypes,
  RotationTypes,
  SignalDetectionTypes,
  StationTypes,
  TimeTypes,
  TypeUtil,
  UserProfileTypes,
  WaveformTypes,
  WaveformUtil,
  WorkflowTypes
};
