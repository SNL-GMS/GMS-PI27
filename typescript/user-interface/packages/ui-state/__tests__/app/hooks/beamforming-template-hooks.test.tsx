import { BeamformingTemplateTypes } from '@gms/common-model';
import {
  asar,
  beamformingTemplatesByBeamTypeByStationByPhase,
  defaultStations,
  eventBeamformingTemplate,
  pdar,
  processingAnalystConfigurationData
} from '@gms/common-model/__tests__/__data__';
import { renderHook } from '@testing-library/react-hooks';
import { produce } from 'immer';

import {
  hasBeamformingTemplatesForQueryArgs,
  useBeamformingTemplatesForEvent,
  useBeamformingTemplatesForFK,
  useBeamformingTemplatesForVisibleStationsAndFavoritePhases
} from '../../../src/ts/app/hooks/beamforming-template-hooks';
import type { AppState } from '../../../src/ts/app/store';
import { getStore } from '../../../src/ts/ui-state';
import { appState, getTestReduxWrapper } from '../../test-util';

jest.mock('../../../src/ts/app/hooks/react-redux-hooks', () => {
  const actualRedux = jest.requireActual('../../../src/ts/app/hooks/react-redux-hooks');
  const mockUseAppDispatch = jest
    .fn()
    .mockImplementation(() => jest.fn(async () => Promise.resolve()));
  return {
    ...actualRedux,
    useAppDispatch: mockUseAppDispatch,
    useAppSelector: jest.fn((stateFunc: (state: AppState) => any) => {
      const state: AppState = produce(appState, draft => {
        draft.app.analyst.phaseSelectorFavorites = {
          'Seismic & Hydroacoustic': [
            'P',
            'Pn',
            'Pg',
            'pP',
            'S',
            'PKP',
            'PKPdf',
            'PKPbc',
            'PKPab',
            'PcP',
            'ScP',
            'Sn',
            'Lg',
            'Rg',
            'sP'
          ]
        };
        draft.data.beamformingTemplates = beamformingTemplatesByBeamTypeByStationByPhase;
      });
      return stateFunc(state);
    })
  };
});

jest.mock(
  '../../../src/ts/app/api/processing-configuration/processing-configuration-api-slice',
  () => {
    const actual = jest.requireActual(
      '../../../src/ts/app/api/processing-configuration/processing-configuration-api-slice'
    );
    return {
      ...actual,
      useGetProcessingAnalystConfigurationQuery: jest.fn(() => ({
        data: processingAnalystConfigurationData
      }))
    };
  }
);

jest.mock('../../../src/ts/app/hooks/station-definition-hooks', () => {
  const actual = jest.requireActual('../../../src/ts/app/hooks/station-definition-hooks');
  return {
    ...actual,
    useVisibleStations: jest.fn(() => defaultStations)
  };
});

describe('test hooks', () => {
  it('is defined', () => {
    expect(useBeamformingTemplatesForVisibleStationsAndFavoritePhases).toBeDefined();
    expect(useBeamformingTemplatesForFK).toBeDefined();
    expect(hasBeamformingTemplatesForQueryArgs).toBeDefined();
  });

  it('gets an FK template', () => {
    const { result } = renderHook(useBeamformingTemplatesForFK, {
      wrapper: getTestReduxWrapper(getStore())
    });
    expect(result.current.data).toMatchSnapshot();
  });

  it('gets an EVENT template', () => {
    const { result } = renderHook(() => useBeamformingTemplatesForEvent([]), {
      wrapper: getTestReduxWrapper(getStore())
    });
    expect(result.current.data).toMatchSnapshot();
  });

  it('hasBeamformingTemplatesForQueryArgs checks a data object to see if the data already exists', () => {
    // returns false if beam type is not requested
    expect(
      hasBeamformingTemplatesForQueryArgs(
        {},
        {
          beamType: BeamformingTemplateTypes.BeamType.EVENT,
          phases: ['P'],
          stations: [pdar]
        }
      )
    ).toBeFalsy();
    // returns false if a station has not been requested
    expect(
      hasBeamformingTemplatesForQueryArgs(
        { [BeamformingTemplateTypes.BeamType.EVENT]: { [pdar.name]: {} } },
        {
          beamType: BeamformingTemplateTypes.BeamType.EVENT,
          phases: ['P', 'S'],
          stations: [pdar, asar]
        }
      )
    ).toBeFalsy();

    // returns false if a phase has not been requested
    expect(
      hasBeamformingTemplatesForQueryArgs(
        {
          [BeamformingTemplateTypes.BeamType.EVENT]: {
            [pdar.name]: { P: eventBeamformingTemplate },
            [asar.name]: { P: eventBeamformingTemplate }
          }
        },
        {
          beamType: BeamformingTemplateTypes.BeamType.EVENT,
          phases: ['P', 'S'],
          stations: [pdar, asar]
        }
      )
    ).toBeFalsy();

    // returns true if all phases and stations have been requested

    expect(
      hasBeamformingTemplatesForQueryArgs(
        {
          [BeamformingTemplateTypes.BeamType.EVENT]: {
            [pdar.name]: { P: eventBeamformingTemplate, S: eventBeamformingTemplate },
            [asar.name]: { P: eventBeamformingTemplate, S: eventBeamformingTemplate }
          }
        },
        {
          beamType: BeamformingTemplateTypes.BeamType.EVENT,
          phases: ['P', 'S'],
          stations: [pdar, asar]
        }
      )
    ).toBeTruthy();
  });
});
