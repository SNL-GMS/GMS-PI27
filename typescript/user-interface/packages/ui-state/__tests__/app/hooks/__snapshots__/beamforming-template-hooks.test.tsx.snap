// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test hooks gets an EVENT template 1`] = `
{
  "PDAR": {
    "P": {
      "beamDescription": {
        "beamSummation": "COHERENT",
        "beamType": "EVENT",
        "phase": "P",
        "preFilterDefinition": undefined,
        "samplingType": "NEAREST_SAMPLE",
        "twoDimensional": true,
      },
      "beamDuration": 300,
      "inputChannels": [
        {
          "effectiveAt": 1636503404,
          "name": "PDAR.PD01.SHZ",
        },
        {
          "effectiveAt": 1636503404,
          "name": "PDAR.PD02.SHZ",
        },
        {
          "effectiveAt": 1636503404,
          "name": "PDAR.PD03.SHZ",
        },
      ],
      "leadDuration": 5,
      "minWaveformsToBeam": 2,
      "orientationAngleToleranceDeg": 5,
      "sampleRateToleranceHz": 0.5,
      "station": {
        "effectiveAt": 1689026400,
        "name": "PDAR",
      },
    },
  },
}
`;

exports[`test hooks gets an FK template 1`] = `
{
  "PDAR": {
    "P": {
      "beamDescription": {
        "beamSummation": "COHERENT",
        "beamType": "FK",
        "phase": "P",
        "preFilterDefinition": undefined,
        "samplingType": "NEAREST_SAMPLE",
        "twoDimensional": true,
      },
      "beamDuration": 300,
      "inputChannels": [
        {
          "effectiveAt": 1636503404,
          "name": "ASAR.AS01.SHZ",
        },
        {
          "effectiveAt": 1636503404,
          "name": "ASAR.AS02.SHZ",
        },
        {
          "effectiveAt": 1636503404,
          "name": "ASAR.AS03.SHZ",
        },
      ],
      "leadDuration": 5,
      "minWaveformsToBeam": 2,
      "orientationAngleToleranceDeg": 5,
      "sampleRateToleranceHz": 0.5,
      "station": {
        "effectiveAt": 1689026400,
        "name": "ASAR",
      },
    },
  },
}
`;
